gbn/timeout_manager.go

```diff
@@ -0,0 +1,338 @@
+		resendTimeout:          defaultResendTimeout,
+		handshakeTimeout:       defaultHandshakeTimeout,
+		useStaticTimeout:       false,
+		resendMultiplier:       defaultResendMultiplier,
+		finSendTimeout:         defaultFinSendTimeout,
+		recvTimeout:            DefaultRecvTimeout,
+		sendTimeout:            DefaultSendTimeout,
+		sentTimes:              make(map[uint8]time.Time),
+		timeoutUpdateFrequency: defaultTimeoutUpdateFrequency,
+	}
+
+	for _, opt := range timeoutOpts {
+		opt(m)
+	}
+
+	return m
+}
+
+// Sent should be called when a message is sent by the connection. The resent
+// parameter should be set to true if the message is a resent message.
+func (m *TimeoutManager) Sent(msg Message, resent bool) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	// We will dynamically update the resend timeout throughout the lifetime
+	// of the connection, to ensure that it reflects the current response
+	// time. Therefore, we'll keep track of when we sent a package, and when
+	// we receive the corresponding response.
+	// If we're resending a message, we can't know if a corresponding
+	// response is the response to the resent message, or the original
+	// message. Therefore, we never update the resend timeout after
+	// resending a message.
+	switch msg := msg.(type) {
+	case *PacketSYN:
+		m.latestSentSYNTimeMu.Lock()
+		defer m.latestSentSYNTimeMu.Unlock()
+
+		if !resent {
+			m.latestSentSYNTime = time.Now()
+
+			return
+		}
+
+		// If we've resent the SYN, we'll reset the latestSentSYNTime to
+		// the zero value, to ensure that we don't update the resend
+		// timeout based on the corresponding response, as we can't know
+		// if the response is for the resent SYN or the original SYN.
+		m.latestSentSYNTime = time.Time{}
+
+	case *PacketData:
+		m.sentTimesMu.Lock()
+		defer m.sentTimesMu.Unlock()
+
+		if resent {
+			// If we're resending a data packet, we'll delete the
+			// sent time for the sequence, to ensure that we won't
+			// update the resend timeout when we receive the
+			// corresponding response.
+			delete(m.sentTimes, msg.Seq)
+
+			return
+		}
+
+		m.sentTimes[msg.Seq] = time.Now()
+	}
+}
+
+// Received should be called when a message is received by the connection.
+func (m *TimeoutManager) Received(msg Message) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	switch msg := msg.(type) {
+	case *PacketSYN, *PacketSYNACK:
+		m.latestSentSYNTimeMu.Lock()
+
+		if m.latestSentSYNTime.IsZero() {
+			m.latestSentSYNTimeMu.Unlock()
+
+			return
+		}
+
+		responseTime := time.Since(m.latestSentSYNTime)
+
+		m.latestSentSYNTime = time.Time{}
+		m.latestSentSYNTimeMu.Unlock()
+
+		m.updateResendTimeout(responseTime)
+
+	case *PacketACK:
+		m.sentTimesMu.Lock()
+
+		sentTime, ok := m.sentTimes[msg.Seq]
+		if !ok {
+			m.sentTimesMu.Unlock()
+
+			return
+		}
+
+		delete(m.sentTimes, msg.Seq)
+
+		m.sentTimesMu.Unlock()
+
+		m.responseCounter++
+
+		reachedFrequency := m.responseCounter%
+			m.timeoutUpdateFrequency == 0
```

Comment:
isn't there a possible race on access to `responseCounter` here? nothing is stopping Receive from being called concurrently.

Reply:
should the main timeoutMgr's mutex not be held for the duration of Receive and not only when updating the resend timeout?

Reply:
Thanks, good idea!

So currently, since the `GoBackNConn` `sendPacketsForever` & `receivePacketsForever` are individually run synchronously, but concurrent from each other, we will never get multiple concurrent `TimeoutManager` `Sent` & `Received` calls, they can only happen concurrent from each other. But I updated to protect the `responseCounter` by the `TimeoutManager` `mu` regardless, as in case the above would ever change in the future :)!

> should the main timeoutMgr's mutex not be held for the duration of Receive and not only when updating the resend timeout?

I actually don't really agree with this though. The reason for that is that in case the `updateResendTimeout` (which also accesses the same `mu`) ever get's more complicated & slow in the future, the measurement of the time it took between a `Sent` & `Received` message might get skewed in the future, as `Received` might then get stuck at accessing the `TimeoutManager` `mu` for an extended time.

Reply:
> are individually run synchronously, but concurr

Yeah we should not depend on the API being used in a synchronous way. 

do we really ever expect `updateResendTimeout` to take long though? 

> he measurement of the time it took between a Sent & Received message might get skewed in the future, as Received might then get stuck at accessing the TimeoutManager mu for an extended time.

I mean, you could just do the `time.Now()` before acquiring the lock so that the time you get the message is still accurate. You could also write a benchmark test to measure the performance of the current impl vs one with just a single mutex to see if it really does make a big difference if you wanted to verify this btw.  [EDIT: Ok I wrote one. see below]

Currently: the `mu` lock is held here and then immediately acquired again in `updateResendTimeout` which is a red flag for a race condition. 

Also just in general having all these locks locked/unlocked at different times without a single long running lock is a bit of a flag for a race condition. Like above we now have this: 

```
		m.sentTimesMu.Unlock()

		m.mu.Lock()
```

which can cause a race if `Received` were to be called by 2 threads at the same time. 

same with the above case: 
```
m.latestSentSYNTimeMu.Unlock()

m.updateResendTimeout(responseTime)
```

Just for fun, i wrote some basic benchmark tests to compare things. So yes, if we must `mu` everywhere for everything, things are a bit slower (around 80ns/op slower). So cool - let's keep the individual mutexes. But let's fix the race's. When I fixed the races the
 benchmark tests show no performance difference. See the patch here: [timeoutmgr.patch](https://github.com/lightninglabs/lightning-node-connect/files/13950552/timeoutmgr.patch)
I've left the benchmark tests there so you can see what they look like (but feel free to delete. Might be worth keeping the second one just for quick testing of any future modifications)

Reply:
Thanks a lot for the great benchmarks and the patch @ellemouton! I've applied it and added it with slight modifications :)!

> do we really ever expect updateResendTimeout to take long though?

No not currently at least. What I tried to say is that it shouldn't effect the measurement of the time it took between a `Sent` & `Received` message even if it were to change in the future. Your latest patch addresses that potential issue though :)!

> you could just do the time.Now() before acquiring the lock

Good idea, thanks :tada:!

> Also just in general having all these locks locked/unlocked at different times without a single long running lock is a bit of a flag for a race condition. Like above we now have this:

Oh I actually thought it was the opposite, TIL then :)! I thought you should avoid holding locks while when acquiring other locks, unless necessary, as that creates lock orders, which then makes it very easy to introduce deadlocks in the future if your not careful.
In terms of the approach when releasing locks before acquiring new ones potentially leading to different threads acquiring the locks before the other threads that attempt to acquire the same lock, I thought that was what's expected in concurrent programming. It could also potentially lead to more optimized asynchronous execution if designed correctly. As in a case where you have a function `func1` that acquires `lock1` followed by `lock2`, if the `lock1` is held while acquiring `lock2` and then releases `lock1`, if a thread has executed `func1` and is temporarily stuck at waiting to acquire `lock2`, any concurrent thread that executes `func1` will now need to wait for the first thread to acquire `lock2` before the code that requires `lock1` can be executed, potentially slowing the overall execution.

I've changed these cases to what your suggestion, but in case my old understanding is valid, let me know and I'll change back :)!

> i wrote some basic benchmark tests to compare things

Awesome, thanks a lot :pray::rocket:!! I kept both of them, as I think it adds extra coverage and it doesn't hurt to have both a synchronous and a concurrent version. Let me know if you really think I should remove the first one though!

Reply:
Your understanding is 100% correct but it just means that we should make sure that the locks are not acquired an a different order else-where (which would lead to a deadlock) and that they are not held for very long. The other option is just to have a single mutex and use that for everything but as you've pointed out, that would slow things down a bit. So if we are going with multiple mutexes then that means we need to just be very careful _but_ it still needs to be correct behaviour. Releasing one lock before acquiring another can lead to undefined behaviour if the  API is used in a multithreaded way.

Reply:
>  I kept both of them, as I think it adds extra coverage and it d
Sounds good!